kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/callbacks/AbstractCallbackRegistryPunctuator.java

```diff
@@ -0,0 +1,69 @@
+package org.hypertrace.core.kafkastreams.framework.callbacks;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.Punctuator;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.hypertrace.core.kafkastreams.framework.callbacks.action.CallbackAction;
+
+public abstract class AbstractCallbackRegistryPunctuator<T> implements Punctuator {
+  private final KeyValueStore<Long, List<T>> objectStore;
+  private final CallbackRegistryPunctuatorConfig config;
+
+  public AbstractCallbackRegistryPunctuator(
+      CallbackRegistryPunctuatorConfig config, KeyValueStore<Long, List<T>> objectStore) {
+    this.config = config;
+    this.objectStore = objectStore;
+  }
+
+  public void invoke(long atTimestampInMs, T object) {
+    long windowAlignedTimestamp = getWindowAlignedTimestamp(atTimestampInMs);
+    List<T> objectsAtWindow =
+        Optional.ofNullable(objectStore.get(windowAlignedTimestamp)).orElse(new ArrayList<>());
+    objectsAtWindow.add(object);
+    objectStore.put(windowAlignedTimestamp, objectsAtWindow);
+  }
+
+  public void cancelInvocation(long atTimestampInMs, T object) {
+    long windowAlignedTimestamp = getWindowAlignedTimestamp(atTimestampInMs);
+    List<T> objectsAtWindow =
+        Optional.ofNullable(objectStore.get(windowAlignedTimestamp)).orElse(new ArrayList<>());
+    objectsAtWindow.remove(object);
+    if (!objectsAtWindow.isEmpty()) {
```

Comment:
Why do we need to put this if it's not empty? Is `get` on a kafka KV store a mutation (a la pop/offer)? If it's a strict read, doesn't seem necessary unless I'm missing something

Reply:
I don't think it is a mutation. We follow this pattern everywhere. RocksDB is interfaced via JNI, and it only understands bytes. Serializers stay one level above and convert everything to bytes.

Now if we just mutate the list, it just modifies in memory reference. Doesn't serialize again and persist in DB, doesn't write to change log topic. Hence every modification must contain a corresponding put. 

I vaguely recollect reading some doc string related to this some where but can't seem to find it at the moment, will share if found.

Reply:
Thanks for the explanation. If that's the case, let's say the list contains one and we remove it. Currently we would skip the put, but don't we need to make sure that gets written back too?

Also if we're effectively working with it like it's immutable - would it make more sense to just use immutable lists so we don't mess this up?

Reply:
> If that's the case, let's say the list contains one and we remove it. Currently we would skip the put, but don't we need to make sure that gets written back too?

Ah my bad, I missed to add delete key in else. Good catch!

> Also if we're effectively working with it like it's immutable - would it make more sense to just use immutable lists so we don't mess this up?

Right I had the same thought, but I'm not sure if we have immutable concrete class we can provide for List Serializer like here?

https://github.com/hypertrace/hypertrace-ingester/blob/59eaa47df3426199c56d5ab8ab4baba4dc29fa0d/raw-spans-grouper/raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansGrouper.java#L78-L83

Reply:
I just checked src of deserializer, it uses `add` method, so may be we need mutable for sake of that? I've modified to take in store of `ArrayList` type as well just to not hide this mutable requirement from interface
```java 
// org.apache.kafka.common.serialization.ListDeserializer#deserialize
List<Inner> deserializedList = createListInstance(size);
            for (int i = 0; i < size; i++) {
                int entrySize = serStrategy == SerializationStrategy.CONSTANT_SIZE ? primitiveSize : dis.readInt();
                if (entrySize == ListSerde.NULL_ENTRY_VALUE || (nullIndexList != null && nullIndexList.contains(i))) {
                    deserializedList.add(null);
                    continue;
                }
                byte[] payload = new byte[entrySize];
                if (dis.read(payload) == -1) {
                    log.error("Ran out of bytes in serialized list");
                    log.trace("Deserialized list so far: {}", deserializedList); // avoid logging actual data above TRACE level since it may contain sensitive information
                    throw new SerializationException("End of the stream was reached prematurely");
                }
                deserializedList.add(inner.deserialize(topic, payload));
            }
```

Reply:
Yeah was just about to say - if this is used by other code we should not make any immutability assumptions.